noir/noir-repo/acvm-repo/acir/src/circuit/opcodes.rs

```diff
@@ -29,6 +29,17 @@ pub enum Opcode {
         block_id: BlockId,
         init: Vec<Witness>,
     },
+    /// Calls to functions represented as a separate circuit. A call opcode allows us
+    /// to build a call stack when executing the outer-most circuit.
+    Call {
+        /// Id for the function being called. It is the responsibility of the executor
+        /// to fetch the appropriate circuit from this id.
+        id: u32,
+        /// Inputs to the function call
+        inputs: Vec<Witness>,
```

Comment:
They made sense to me more as pure witnesses for a few reasons.
1. The inputs to a single ACIR circuit are going to be witnesses (thinking about the `Circuit` struct in this instance)
2. The `Call` opcode is similar to a black box function in that we are asking the executor/backend to treat this opcode differently depending upon the implementation. For black box functions we expect pure witnesses rather than expressions. 
3. We mainly use expressions to evaluate the result of some polynomial for either asserting zero or checking whether some predicate is zero or one. While for a witness I just want to fetch some value as the arguments for this `Call`.

Did you have something in mind with leaving these as expressions?

Reply:
> The inputs to a single ACIR circuit are going to be witnesses (thinking about the Circuit struct in this instance)

The ACVM can evaluate these expressions to generate the input witnesses to the called acir circuit much in the same way we're currently doing for brillig calls however.

During folding however we're going to need to relate the inputs specified in the `Call` opcode to the inputs which were passed to the folded functions. I'm not 100% clear on if this process this may cause a hurdle to passing in expressions. In the pessimistic case where we require the inputs to be witnesses however, (e.g. the backend is inlining the called acir) it could inject constraints for these expressions to create the input witnesses for the called acir.

> The Call opcode is similar to a black box function in that we are asking the executor/backend to treat this opcode differently depending upon the implementation. For black box functions we expect pure witnesses rather than expressions

I'm just thinking about this as similar to brillig calls where we allow the inputs to be expressions (or even memory arrays). If we could maintain this then we could reuse the same mechanism for calling acir functions for calling brillig functions whereas if we need to pass witnesses then we're going be adding a lot of unnecessary constraints.

Reply:
> I'm just thinking about this as similar to brillig calls where we allow the inputs to be expressions (or even memory arrays). If we could maintain this then we could reuse the same mechanism for calling acir functions for calling brillig functions whereas if we need to pass witnesses then we're going be adding a lot of unnecessary constraints.

Ok I see your point here. Worst case, the backend can inject the extra constraints it needs.

Reply:
I do think these constraints are going to end up being written somewhere though, either by the compiler or by the backend though. In Brillig we didn't have to constrain the input as it is an unconstrained environment.

Reply:
> I do think these constraints are going to end up being written somewhere though, either by the compiler or by the backend though. In Brillig we didn't have to constrain the input as it is an unconstrained environment.

Due to this I do lean towards that it is better to have the compiler lay down the constraints vs. the backend.

Reply:
Leaving as witnesses also lets us keep the interface for Calls cleaner and it is not clear that using expressions will actually allow us to avoid unnecessary constraints. If it does enable us to remove unnecessary constraints, we can update and optimize later imo.

Reply:
yep, I'm fine leaving it as witnesses for the first cut as well.

Reply:
Should we merge then?